### PR TITLE
Add feature to ban the participants from participating in a challenge

### PR DESCRIPTION
**Pull request by Sanji515**
#### Changes proposed in this pull request:

- Added one more new tab named `Settings` on the challenge page for the host user

- Listed all the banned email ids with pencil edit icon. On clicking it, a text box with the all the banned email ids as it's value will be shown.

- When editing there will be two options (icons), one to save the changes and another one (cross icon) to discard the changes

- Link to live demo: http://pr-163-evalai.surge.sh

**<p align="center">Screenshot 1:</p>**

Screenshot from 2019-07-28 21-02-58

**<p align="center">Screenshot 2:</p>**

Screenshot from 2019-07-28 21-03-48

### CODE REVIEW

**Comment by RishabhJain2018** [comment]
@Sanji515 Please resolve the conflicts here.

**Comment by lunayach** [comment]
@Sanji515 resolve the conflicts here.

**Comment by Sanji515** [comment]
@lunayach @galipremsagar @Shekharrajak Please review

**Comment by galipremsagar** [comment]
@Sanji515 resolve conflicts?

**Comment by Sanji515** [comment]
@galipremsagar I've resolved the conflicts, please review.

File: src/app/components/challenge/challengesettings/challengesettings.component.html [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                  </i>
+                </a>
+              </span>
+            </span>
+          </span>
+        </div>
+      </div>
+    </div>
+  </div>
+</div>
```
Comment (galipremsagar) [inline src/app/components/challenge/challengesettings/challengesettings.component.html]: Add a new-line at the end of all newly added files.
```suggestion
</div>

```

**Review by galipremsagar — CHANGES_REQUESTED** [review]

File: src/app/components/challenge/challengesettings/challengesettings.component.html (hunk 1 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.html]: @galipremsagar done

File: src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  updateView() {
+    this.bannedEmailIds = this.challenge.banned_email_ids;
+  }
+
+  getBannedEmailsListObject(bannedEmailInputValue) {
+    const bannedEmailIds = [];
+    // First remove all the whitespaces and then split
+    const splittedEmails = bannedEmailInputValue.replace(/\s/g, '').split(',');
```
Comment (galipremsagar) [inline src/app/components/challenge/challengesettings/challengesettings.component.ts]: `replace(/\s/g, '')`
Why this replace?

**Review by galipremsagar — CHANGES_REQUESTED** [review]
If the input is empty and we click on green tick it doesn't accept. Either we should we showing some error or just accept the input and process it.

File: src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 2 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.ts]: @galipremsagar to remove all the whitespaces like suppose the list is like `abc@gmail.com, xyz@gmail.com` so I don't want the space after comma `,`.

Comment (galipremsagar) [reply src/app/components/challenge/challengesettings/challengesettings.component.ts]: we can do this: 

```suggestion
    const splittedEmails = bannedEmailInputValue.split(",").map(item => item.trim());
```

File: src/app/components/challenge/challengesettings/challengesettings.component.html [hunk 3]
```diff
@@ -0,0 +1,55 @@
+
```
Comment (lunayach) [inline src/app/components/challenge/challengesettings/challengesettings.component.html]: Remove this empty line.

File: src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        data => {
+          SELF.challenge.banned_email_ids = data.banned_email_ids;
+          SELF.isBannedEmailInputVisible = false;
+          SELF.globalService.showToast('success', 'Banned participant emails are successfully updated!', 5);
+        },
+        err => {
+          SELF.globalService.handleApiError(err, true);
+          SELF.globalService.showToast('error', err);
+        },
+        () => console.log('BANNED-EMAIL-IDS-UPDATE-FINISHED')
```
Comment (lunayach) [inline src/app/components/challenge/challengesettings/challengesettings.component.ts]: Remove this.

File: src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 2 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.ts]: done

File: src/app/components/challenge/challengesettings/challengesettings.component.html (hunk 3 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.html]: @lunayach done

File: src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 4 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.ts]: @lunayach done

File: src/app/components/challenge/challengesettings/challengesettings.component.html [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+              <a class="pointer li-mr" (click)="isBannedEmailInputVisible=
+                false">
+                <i class="fa fa-times red-text fa-lg" aria-hidden="true"></i>
+              </a>
+            </span>
+          </span>
+          <span *ngIf="!isBannedEmailInputVisible" class="display-inline">
+            <span class="row">
+              <span *ngIf="challenge.banned_email_ids" class="col-md-7 col-sm-7 li-mr">
+                <span style="white-space: pre-line;" *ngFor="let participant_email of challenge.banned_email_ids"
```
Comment (lunayach) [inline src/app/components/challenge/challengesettings/challengesettings.component.html]: Add `style` to `.scss` file.

File: src/app/components/challenge/challengesettings/challengesettings.component.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  updateView() {
+    this.bannedEmailIds = this.challenge.banned_email_ids;
+  }
+
+  getBannedEmailsListObject(bannedEmailInputValue) {
+    const bannedEmailIds = [];
+    // First remove all the whitespaces and then split
+    const splittedEmails = bannedEmailInputValue.split(',').map(item => item.trim());
+    for (let i = 0; i < splittedEmails.length; i++) {
+      bannedEmailIds.push(splittedEmails[i]);
```
Comment (lunayach) [inline src/app/components/challenge/challengesettings/challengesettings.component.ts]: Should we also apply an email `regex` check for each email?

**Comment by lunayach** [comment]
As we add email ID's one by one, should we have something like this (https://stackblitz.com/angular/earboyoxrglg?file=app%2Fchips-input-example.ts). That is, each email is added as a chip. (IDK, but this looks more intuitive)

File: src/app/components/challenge/challengesettings/challengesettings.component.ts (hunk 6 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.ts]: @lunayach done

File: src/app/components/challenge/challengesettings/challengesettings.component.html (hunk 5 above)
Comment (Sanji515) [reply src/app/components/challenge/challengesettings/challengesettings.component.html]: @lunayach Removed it as there's no need of it as now used `chips` to show the banned emails, thanks :)

**Comment by Sanji515** [comment]
> As we add email ID's one by one, should we have something like this (https://stackblitz.com/angular/earboyoxrglg?file=app%2Fchips-input-example.ts). That is, each email is added as a chip. (IDK, but this looks more intuitive)

Yes, it looks great. Implemented it and removed the `tick` and `cross` icon. Please take a look @lunayach @RishabhJain2018

**Review by galipremsagar — APPROVED** [review]
Changes LGTM, This can be merged.

**Comment by galipremsagar** [comment]
@Sanji515  Just update the screenshots of the UI you have posted above as it looks much nicer and better now

**Pull request merged**